Implement TLSSocket

SeppoTakalo opened the pull request:
### Description

Provide TLSSocket implementation. This implementation allows TLSSocket
to wrap around any existing socket. Currently only TLS supported. DTLS
not yet implemented.

Reviewed design document also provided with the implementation.

Documentation submitted to Mbed OS handbook.
https://github.com/ARMmbed/mbed-os-5-docs/pull/759

NOTE: Design review has already been passed earlier. This PR is not requesting API review or design suggestions anymore.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Breaking change

Commit:
Implement TLSSocket

Provide TLSSocket implementation. This implementation allows TLSSocket
to wrap around any existing socket. Currently only TLS supported. DTLS
not yet implemented.

Design document also provided with the implementation.

Documentation submitted to Mbed OS handbook.

SeppoTakalo commented:
@kjbracey-arm Please review.

SeppoTakalo commented:
I'll fix the Doxygen documentation blocks.

SeppoTakalo commented:
@0xc0170 Please test.

0xc0170 commented:
@AnotherButler Please review the design document

@SeppoTakalo Any security review needs to be done (tls team or similar) or all fine here?

SeppoTakalo commented:
@0xc0170 We have done the internal reviews already with TLS team and architecture group. See internal Jira ticket ONME-3788 for a record.

Thanks.

0xc0170 commented:
/morph build

0xc0170 reviewed (approved):
There are 3 TODO left, shall they be captured somewhere?

0xc0170 commented:
Failures look related to the changeset

0xc0170 commented:
/morph build

SeppoTakalo commented:
Can this be retested now?

The test result is showing previous test just, just a minutes before rebase and fixing the Arm compiler issue.

cmonr commented:
/morph build

cmonr commented:
IAR network license issue.

/morph export-build
/morph mbed2-build

cmonr commented:
@ARMmbed/mbed-os-test When y'all have a moment, unsure what's going on with the cloud-client-test

cmonr commented:
Error located: 
> [Build K66F ARM] "/tmp/Mx9GAo", line 120 (column 3): Warning: L6312W: Empty Execution region description for region RW_IRAM1
[Build K66F ARM] Error: L6218E: Undefined symbol mbedtls_strerror (referred from BUILD/K66F/ARM/mbed-os/features/netsocket/TLSSocketWrapper.o).
[Build K66F ARM] Finished: 0 information, 1 warning and 1 error messages.
[Build K66F ARM] [ERROR] "/tmp/Mx9GAo", line 120 (column 3): Warning: L6312W: Empty Execution region description for region RW_IRAM1
[Build K66F ARM] Error: L6218E: Undefined symbol mbedtls_strerror (referred from BUILD/K66F/ARM/mbed-os/features/netsocket/TLSSocketWrapper.o).
[Build K66F ARM] Finished: 0 information, 1 warning and 1 error messages.
[Build K66F ARM] 
[Build K66F ARM] [mbed] ERROR: "/usr/bin/python" returned error.
[Build K66F ARM]        Code: 1
[Build K66F ARM]        Path: "/builds/ws/mbed-os-ci_cloud-client-test/mbed-cloud-client-example"
[Build K66F ARM]        Command: "/usr/bin/python -u /builds/ws/mbed-os-ci_cloud-client-test/mbed-cloud-client-example/mbed-os/tools/make.py -t ARM -m K66F --source . --build ./BUILD/K66F/ARM"
[Build K66F ARM]        Tip: You could retry the last command with "-v" flag for verbose output
[Build K66F ARM] ---
[Build K66F ARM] [mbed] Auto-installing missing Python modules...

@ARMmbed/mbed-os-test The build step should have probably failed.

Commit:
Check that MBEDTLS_ERROR_C is defined before using mbedtls_strerror().

If MBEDTLS_ERROR_C is not defined, that strerror function does not
exist, or is just dummy definition if MBEDTLS_ERROR_STRERROR_DUMMY is
defined.

SeppoTakalo commented:
Cloud client build now fixed.

cmonr commented:
/morph build
